pkg/bloomcompactor/bloomcompactor.go

```diff
@@ -463,92 +463,95 @@ func (c *Compactor) runCompact(ctx context.Context, logger log.Logger, job Job,
+		if err := os.RemoveAll(localDst); err != nil {
+			level.Error(logger).Log("msg", "failed to remove block directory", "dir", localDst, "err", err)
+		}
+	}()
+
+	var resultingBlock bloomshipper.Block
+	defer func() {
+		if resultingBlock.Data != nil {
+			_ = resultingBlock.Data.Close()
+		}
+	}()
+
+	if len(blocksMatchingJob) == 0 && len(metasMatchingJob) > 0 {
+		// There is no change to any blocks, no compaction needed
+		level.Info(logger).Log("msg", "No changes to tsdb, no compaction needed")
+		return nil
+	} else if len(metasMatchingJob) == 0 {
+		// No matching existing blocks for this job, compact all series from scratch
 
-		blockOptions := v1.NewBlockOptions(bt.GetNGramLength(), bt.GetNGramSkip())
 		builder, err := NewPersistentBlockBuilder(localDst, blockOptions)
 		if err != nil {
 			level.Error(logger).Log("msg", "creating block builder", "err", err)
 			return err
 		}
 
 		fpRate := c.limits.BloomFalsePositiveRate(job.tenantID)
-		storedBlock, err := compactNewChunks(ctx, logger, job, fpRate, bt, storeClient.chunk, builder)
+		resultingBlock, err = compactNewChunks(ctx, logger, job, fpRate, bt, storeClient.chunk, builder)
 		if err != nil {
 			return level.Error(logger).Log("msg", "failed to compact new chunks", "err", err)
 		}
 
-		archivePath := filepath.Join(c.cfg.WorkingDirectory, uuid.New().String())
+	} else if len(blocksMatchingJob) > 0 {
+		// When already compacted metas exists, we need to merge all blocks with amending blooms with new series
 
-		blockToUpload, err := c.compressBloomBlock(storedBlock, archivePath, localDst, logger)
+		resultingBlock, err = mergeCompactChunks(ctx, logger, c.bloomShipperClient, storeClient, bt, job, blockOptions, blocksMatchingJob, c.cfg.WorkingDirectory, localDst)
 		if err != nil {
-			level.Error(logger).Log("msg", "putting blocks to storage", "err", err)
+			level.Error(logger).Log("msg", "failed to merge existing blocks with new chunks", "err", err)
 			return err
 		}
-		defer func() {
-			err = os.Remove(archivePath)
-			if err != nil {
-				level.Error(logger).Log("msg", "removing archive file", "err", err, "file", archivePath)
-			}
-		}()
+	}
 
-		// Do not change the signature of PutBlocks yet.
-		// Once block size is limited potentially, compactNewChunks will return multiple blocks, hence a list is appropriate.
-		storedBlocks, err := c.bloomShipperClient.PutBlocks(ctx, []bloomshipper.Block{blockToUpload})
+	archivePath := filepath.Join(c.cfg.WorkingDirectory, uuid.New().String())
+
+	blockToUpload, err := bloomshipper.CompressBloomBlock(resultingBlock.BlockRef, archivePath, localDst, logger)
+	if err != nil {
+		level.Error(logger).Log("msg", "compressing bloom blocks in to tar files", "err", err)
+		return err
+	}
+	defer func() {
+		err = os.Remove(archivePath)
 		if err != nil {
-			level.Error(logger).Log("msg", "putting blocks to storage", "err", err)
-			return err
+			level.Error(logger).Log("msg", "removing archive file", "err", err, "file", archivePath)
```

Comment:
I would write messages of logs that are only logged in case of an error in negative form, such as:
`failed to ...`, or `did not ...`

Reply:
i think I made a wrong observation previously in another components, bc I was only writing like this modelling how we do elsewhere. I'll review all and fix the errors with a negative phrase.